ext/js/lib/js/browser.rb

```diff
@@ -0,0 +1,55 @@
+require "singleton"
+require_relative "./url_resolver"
+
+module JS
+  ScriptLocation = Data.define(:url, :filename)
+
+  class Browser
```

Comment:
I feel that the name "browser" represents a role that is a bit too broad. I am trying to think of a better name.